components/automate-cli/cmd/chef-automate/cmdUtils.go

```diff
@@ -0,0 +1,428 @@
+// RemoteJobs function will run all remote jobs.
+func (c *CmdUtilImpl) RemoteJobs(command string, inputFiles map[string]string, outputFiles []string, remoteService string, errorCheck bool, newSSHUtil SSHUtil, resultChan chan CmdResult) {
+	rc := CmdResult{newSSHUtil.getSSHConfig().hostIP, []string{}, "", nil}
+	if len(inputFiles) != 0 {
+		for inputFile, outputFile := range inputFiles {
+			err := newSSHUtil.copyFileToRemote(inputFile, outputFile, false)
+			if err != nil {
+				rc.Error = err
+				resultChan <- rc
+				return
+			}
+		}
+	}
+
+	output, err := newSSHUtil.connectAndExecuteCommandOnRemote(command, true)
+	if err != nil {
+		rc.Error = err
+		resultChan <- rc
+		return
+	}
+
+	if len(outputFiles) != 0 {
+		for _, file := range outputFiles {
+			outFile := newSSHUtil.getSSHConfig().hostIP + "_" + file
+			destFileName, err := newSSHUtil.copyFileFromRemote(file, outFile)
+			if err != nil {
+				rc.Error = err
+				resultChan <- rc
+				return
+			}
+			writer.Printf("File downloaded %s \n", destFileName)
+			rc.OutputFiles = append(rc.OutputFiles, destFileName)
+		}
+	}
+
+	if errorCheck {
+		err = checkResultOutputForError(output)
+		if err != nil {
+			rc.Error = err
+			resultChan <- rc
+			return
+		}
+	}
+
+	rc.Output = output
+	resultChan <- rc
+}
+
+// preCmdExecCheck will check and execute PreExec function
+func preCmdExecCheck(node *Cmd, sshUtil SSHUtil, infra *AutomteHAInfraDetails, remoteService string, timestamp string, writer *cli.Writer) ([]string, error) {
+	var nodeIps []string
+	var err error
+	if (node.PreExec) != nil {
+		err = node.PreExec(node.CmdInputs, sshUtil, infra, remoteService, timestamp, writer)
+		if err != nil {
+			return nodeIps, err
+		}
+	}
+
+	if remoteService == CONST_FRONTEND {
+		nodeIps, err = isFrontendIPs(node.CmdInputs.NodeIp, infra)
+	} else if remoteService == CONST_AUTOMATE {
+		nodeIps, err = isAutomateIPs(node.CmdInputs.Single, node.CmdInputs.NodeIp, infra)
+	} else if remoteService == CONST_CHEF_SERVER {
+		nodeIps, err = isChefserverIPs(node.CmdInputs.Single, node.CmdInputs.NodeIp, infra)
+	} else if remoteService == CONST_POSTGRESQL {
+		nodeIps, err = isPostgresqlIPs(node.CmdInputs.Single, node.CmdInputs.NodeIp, infra)
+	} else if remoteService == CONST_OPENSEARCH {
+		nodeIps, err = isOpensearchIPs(node.CmdInputs.Single, node.CmdInputs.NodeIp, infra)
+	}
+
+	if err != nil {
+		return nodeIps, err
+	}
+	if len(nodeIps) == 0 {
+		errMsg := fmt.Sprintf("No %s IPs are found", remoteService)
+		return nodeIps, errors.New(errMsg)
+	}
+	return nodeIps, nil
+}
+
+// getSshDetails will return the SSH details.
+func getSshDetails(infra *AutomteHAInfraDetails) *SSHConfig {
+	sshConfig := &SSHConfig{
+		sshUser:    infra.Outputs.SSHUser.Value,
+		sshPort:    infra.Outputs.SSHPort.Value,
+		sshKeyFile: infra.Outputs.SSHKeyFile.Value,
+	}
+	return sshConfig
+}
+
+// isFrontendIPs will return the Frontend Node Ips.
+func isFrontendIPs(ip string, infra *AutomteHAInfraDetails) ([]string, error) {
```

Comment:
This function is used for two purposes. One is to get all IPs of the service. And Second is to check whether the given IP belongs to the service.
Please break it into two functions. One could be getFrontedIPs which only returns all fronted IPs. And Second could be isFrontendIPs which will internally call getFrontedIPs to get the list of IPs and then validate against it.

Reply:
Discussed